### ISSUE TITLE
How to apply cosine similarity loss function in unsupervised training?

### ISSUE THREAD

**Issue by Ceceu** [tensorflow/tensorflow#37882]
In the Figure below is showed a simple deep learning architecture capable of learning embeddings for sentences. The training set is composed by sentence pairs `[[sentence_a],[sentence_b]]` that have the same semantics.

[![enter image description here][1]][1]

The objective is to fine-tune the embeddings of the sentences to be similar (since sentences in the pair have the same semantics). Consequently, a possible loss function would be `CosineSimilarity` loss.

`Encoder 1` and `Encoder 2` have the same definition:

```python
from tensorflow.keras import layers


class LSTMEncoder(layers.Layer):

    def __init__(self,
                 units,  # dimensionality of the output space
                 input_dim,  # vocab of size
                 output_dim,  # embedding dimension
                 name='encoder',
                 **kwargs):
        super(Encoder, self).__init__(name=name, **kwargs)
        self.embedding = layers.Embedding(input_dim=input_dim, output_dim=output_dim)
        self.lstm = layers.LSTM(units=units)

    def call(self, inputs):
        emb = self.embedding(inputs)
        return self.lstm(emb)
```

And the SEntence Representations model `(SERModel)` looks like:

```python
import tensorflow as tf

from source.layer.rnn_encoder import LSTMEncoder


class SERModel(tf.keras.Model):
    """
    Defines a SEntence Representations Model
    """

    def __init__(self,
                 units,  # dimensionality of the output space
                 input_dim,  # vocab of size
                 output_dim,  # embedding dimension
                 name='SEntence Representations Model',
                 **kwargs):
        super(SERModel, self).__init__(name=name, **kwargs)
        self.encoder = LSTMEncoder(units=units, input_dim=input_dim, output_dim=output_dim)

    def call(self, inputs):
        sentence_a, sentence_b = inputs
        return self.encoder(sentence_a), self.encoder(sentence_b)

```

To train the model one can simply do as shown next:

```python
from source.model.SERModel import SERModel
import tensorflow as tf


def run_test():
    
    # load train data
    train_data = load_data()

    ser = SERModel(units=128, input_dim=10000, output_dim=64)

    optimizer = tf.keras.optimizers.Adam(learning_rate=1e-3)

    ser.compile(optimizer, loss=tf.keras.losses.CosineSimilarity())
    ser.fit(train_data, epochs=3, batch_size=64)


if __name__ == '__main__':
    run_test()

```

At this point is where I need help. The loss requires `y_true` and `y_pred`, but as can be seen, this is unsupervised training and there is no `y_true`. So, would you have an approach that would allow me to train the above architecture using `CosineSimilarity` loss?

Thank you very much in advance.

  [1]: https://i.stack.imgur.com/Px8hL.png

**Comment by gowthamkpr** [comment]
@Ceceu Please take a look at the answers of this issue [here](https://www.researchgate.net/post/Loss_Functions_for_unsupervised_image_segmentation).

Also as this issue is not related to bug/performance, build/install, docs or related issues, can you please post this question in stackoverflow as there is a wider community to respond and close this issue. Thanks!

**Comment by gowthamkpr** [comment]
@Ceceu Are you able to figure this issue out? Did you post it in stackoverflow?

**Comment by gowthamkpr** [comment]
Closing this issue as it has been inactive for more than 2 weeks. Please add additional comments for us to open the issue again.
